### DISCUSSION TITLE
A fork with proposals for several new features

### DISCUSSION THREAD

**Discussion by ilg-ul** [facebook/docusaurus#9822] [Ideas]
For the needs of a project (multiple blog instances, with historical data), I had to do a fork and add several enhancements, mainly related to grouping and sorting.

The fork is available at <https://github.com/cronica-it/docusaurus-chronology>.

The updated code is in the `development` branch, while the original code remains in the `main` branch.

The changes, mainly enhancements, are enabled via configuration options, and the defaults should not affect the functionality of the standard Docusaurus project.

The enhancements are listed below.

## Generate grouping pages by Authors, similarly to Tags

For blogs with multiple authors, it is useful to view the pages contributed by each author.

Similarly to the Tags grouping pages, pages to group posts by authors can now be generated.

![authors](https://github.com/facebook/docusaurus/assets/362732/14161677-cd91-4675-ace7-926166b4c549)

Links to authors are also added at the bottom of the posts, after the links to tags:

![authors2](https://github.com/facebook/docusaurus/assets/362732/6d476284-a961-4fe2-8532-fea4cdad4fbf)

A new configuration option `generateAuthorsPages` was added; when set to true, the authors pages are generated as `blog/authors` and `blog/authors/${name}`.

Note: Docusaurus allows to define authors identified by a picture, without having to enter the name; these nameless authors are filtered out and not indexed in the Authors pages.

## Add `showLastUpdateTime` and `showLastUpdateAuthor` to blog posts

In some cases, for example sites with research data, the content of the posts is expected to be updated from time to time, as new data is discovered.

To help users identify the updated content, it is useful to know the last updated time and provide a way to order post by it.

The mechanism of computing the last updated time is based on the latest git commit time, similarly to the mechanism used for docs; it is enabled by the same configuration options, `showLastUpdateTime` and `showLastUpdateAuthor`.

When enabled, the last update time is displayed below the post and also used when sorting the posts in the feed files.

![last-updated](https://github.com/facebook/docusaurus/assets/362732/1b3c90e8-f99d-48bb-b6c5-b27bf968ee19)

## Add `sortSidebarByLastUpdate` to change the order in the sidebar

When posts are edited and the retrieving the date of the last update is enabled, it also makes sense to show the updated posts in the top of the sidebar list.

A new configuration option `sortSidebarByLastUpdate` was added; when set to true, the sidebar list is sorted by the last update time, when available.

## Hide redundant year in dates shown in the Archive page

In the Archive page, when listing the the post titles grouped by years, there is not need to show the year again, since it is redundant.

![hide-year](https://github.com/facebook/docusaurus/assets/362732/39792b83-fe29-4bfa-a009-2e1db8ad2c95)

For aesthetic reasons, a new configuration option `hidePostYearInArchive` was added; when set to true, the date format is adjusted to no longer show the year.

## Improve support for posts related to events in the past

The common use case for blogs is to document recent events; in other words, the post date and the event date are more or less the same.

For blogs documenting historical events, the post date remains in the present, as for regular posts, but the event date is in the past, and should be entered as a separate frontMatter string property.

Since some historical events do not have an exact date, this property can be incomplete, without day or even without month, for example:

```txt
event_date: '1994'
event_date: '1994-11'
event_date: '1994-11-07'
```

For events that lasted more than one day, it is possible to also define the end date, as a string with the similar incomplete syntax:

```txt
event_end_date: '1995'
event_end_date: '1994-12'
event_end_date: '1994-11-08'
```

To enable this feature, a new configuration option `sortPostsByEventDate` was added; when set to true, the event dates are parsed and used when sorting the posts in the Archive page, so the years of past events are located at the end.

![event-year-only](https://github.com/facebook/docusaurus/assets/362732/7937d7f9-8e69-4657-a852-cc64ec8477bf)

## Add `pageBasePath` to blog plugin configuration

In Docusaurus, the URL parts used to compose the pages paths are configurable via options like `routeBasePath`, `tagsBasePath`, etc, but in the path used for multi-page lists, a part of the URL is hard-coded as `page`.

For consistency reasons, a new configuration option was added, `pageBasePath`, allowing to also configure this path, a feature useful for example when all paths are translated to local languages.

## Apply `applyTrailingSlash` to feed URLs

The `trailingSlash` configuration is available to customize the presence/absence of a trailing slash at the end of URLs; unfortunately this setting does not apply to the links generated in the feed files, and tools like Algolia that use the files may behave unusual.

Thus, the `applyTrailingSlash` function is called to conditionally adjust the URLs generated in the feed files.

---

I'd be glad to see these changes contributed back to the upstream project.

So, if there is interest in any of these features, please feel free to use my code as a starting point and possibly improve it.

I can also prepare separate PRs, if needed.

**Comment by slorber** [comment]
Why not create one proposal issue per thing you want to contribute?  

Discussing 7 things in a single issue is only going to create a mess, and not a good idea. 

Otherwise, some of those changes look like possible things we'd want to add to Docusaurus yes.

**Reply by ilg-ul** [reply]
> ... create one proposal issue per thing ...

Issue or canny?

**Reply by slorber** [reply]
If you already have a design proposal + mockup or screenshots, you can create a RFC / Design proposal:

https://github.com/facebook/docusaurus/issues/new?assignees=&labels=feature%2Cstatus%3A+needs+triage&projects=&template=feature.yml

**Comment by ilg-ul** [comment]
Separate issues were added for each new feature.
